vtex/faststore
# chore(gatsby-plugin-cms): Improve cms graphql types generation

## What's the purpose of this pull request?
This PR improves graphql type generation for the cms plugin. 

Currently, to use the CMS, the user has to declare content types using a language called Json Schema. This language is a well defined language and shapes how the data is displayed/stored on the CMS database. Today, we don't use these types for generating the graphql layer of the CMS, but rather create a generic, opaque graphql layer. This layer uses custom `scalar` types and omits important data types, requiring manual typing of the typescript code when consuming this data. This PR tries to solve parts of this problem by using the Gatsby's type inference based on the CMS data. Further work may be required to actually use CMS declared Json schema for creating GraphQL types

## How it works? 
Currently, we use [gatsby-graphql-toolkit](https://www.npmjs.com/package/gatsby-graphql-source-toolkit) for replicating the CMS's graphql layer into the gatsby graphql schema. This generates, in our graphql layer a type like:
```ts
type Block = {
  name: string
  props: any
}

type CmsPageContent = {
  blocks: Block[]
  extraBlocks: Array<{ 
    name: string
    blocks: Block[]
  }>
  beforeBlocks: Block[]
  afterBlocks: Block[]
}
```

As you can see, props has an any type. 

By removing the gatsby-graphql-toolkit and using Gatsby's createNode function, along with Gatsby's automatic type inference, we are able to add types to that `props` property of the type Block.

This is good, however there was another part missing. The extraBlocks option doesn't really need to be an array. Arrays are good when we need to keep an order (like in the case of blocks), however, union type arrays are horrible for typing and consuming, specially on graphql, where you need to create multiple fragments for each case. To solve this problem I have created a key for each block name on the final type, so we can consume it via direct property access.

To summarize all changes and give a final example, let's suppose we have the following contentType written in Json Schema
```
{
  homePage: {
    blocks: {
      banner: {
        type: 'object',
        properties: {
          imageUrl: {
            type: "string",
          }
        }
      }
    },
    extraBlocks: {
      seo: {
        tags: {
          type: 'object'
          properties: {
            title: {
              type: 'string'
            },
            description: {
              type: 'string'
            }
          }
        }
      }
    }
  }
}
```

This will generate the following type on the gatsby's graphql layer
```ts
type Block = {
  name: string
  props: any
}

type CmsHomePage = {
  sections: Block[]
  seo: {
    tags: {
       title: string
       description: string
    }
  }
}
```

And we can query this data on React components by writting the following graphql query:
```gql
query HomePageQuery {
  cmsHomePage {
    sections: {
      name
      props
    }
    seo {
      tags {
        title
        description
      }
    }
  }
}
```

## How to test it?
This is an architectural only change, no feature on the website should be changed. For this, make sure all content on the website stays the same from master

websites:
https://github.com/vtex-sites/btglobal.store/pull/721
https://github.com/vtex-sites/storecomponents.store/pull/1050
https://github.com/vtex-sites/marinbrasil.store/pull/564
